### ISSUE TITLE
Cognito StartWithRefreshTokenAuthAsync incorrectly checks SessionTokens.ExpirationTime, fails with "User is not authenticated."

### ISSUE THREAD

**Issue by jeffreythomasprice** [aws/aws-sdk-net-extensions-cognito#24]
## Expected Behavior

Invoking `StartWithRefreshTokenAuthAsync` on an instance of CognitoUser that had previously authenticated, but now has an expired access token should result in a new access token with an expiration date in the future.

## Current Behavior

`StartWithRefreshTokenAuthAsync` throws the following exception:
```
Amazon.CognitoIdentityProvider.Model.NotAuthorizedException: User is not authenticated.
   at Amazon.Extensions.CognitoAuthentication.CognitoUser.EnsureUserAuthenticated()
   at Amazon.Extensions.CognitoAuthentication.CognitoUser.CreateRefreshTokenAuthRequest(AuthFlowType authFlowType)
   at Amazon.Extensions.CognitoAuthentication.CognitoUser.<StartWithRefreshTokenAuthAsync>d__82.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
... ommitted for brevity
```

## Possible Solution

My workaround right now is to manually assign the user's `SessionTokens` to a new instance created with the existing tokens, but with an expiration time set in the future by an arbitrary amount seems to let it work around the check to see if the tokens are expired.
```
user.SessionTokens = new CognitoUserSession(
	user.SessionTokens.IdToken,
	user.SessionTokens.AccessToken,
	user.SessionTokens.RefreshToken,
	user.SessionTokens.IssuedTime,
	DateTime.Now.AddHours(1)
);
var response = await user.StartWithRefreshTokenAuthAsync(new InitiateRefreshTokenAuthRequest
{
	AuthFlowType = AuthFlowType.REFRESH_TOKEN
});
```

A real solution is probably just to remove the call to `EnsureUserAuthenticated` from the top of `CreateRefreshTokenAuthRequest` in `CognitoUserAuthentication.cs`.
```
private InitiateAuthRequest CreateRefreshTokenAuthRequest(AuthFlowType authFlowType)
{
	// This line is the one that I suggest be removed
	EnsureUserAuthenticated();

	if (authFlowType != AuthFlowType.REFRESH_TOKEN && authFlowType != AuthFlowType.REFRESH_TOKEN_AUTH)
	{
		throw new ArgumentException("authFlowType must be either \"REFRESH_TOKEN\" or \"REFRESH_TOKEN_AUTH\"", "authFlowType");
	}
```

## Steps to Reproduce (for bugs)

Authenticate a cognito user via some variant of the following:
```
AmazonCognitoIdentityProviderClient provider = new AmazonCognitoIdentityProviderClient(new AnonymousAWSCredentials(), RegionEndpoint.USEast2);
CognitoUserPool userPool = new CognitoUserPool(Constants.UserPoolId, Constants.ClientId, provider);
user = new Amazon.Extensions.CognitoAuthentication.CognitoUser(username, Constants.ClientId, userPool, provider);

await user.StartWithSrpAuthAsync(new InitiateSrpAuthRequest()
{
	Password = password
});
```

Wait an hour for the access token to expire.

Attempt to refresh the user:
```
await user.StartWithRefreshTokenAuthAsync(new InitiateRefreshTokenAuthRequest
{
	AuthFlowType = AuthFlowType.REFRESH_TOKEN
});
```

Observe that it throws:
```
Amazon.CognitoIdentityProvider.Model.NotAuthorizedException: User is not authenticated.
```

## Context

I need a cognito user that is as long-lived as it would be on any other client (ex. our web client handles refreshes just fine, allowing multi-hour sessions). If I log in to the desktop client created using this library, and then just leave that client running for an hour all API calls start failing with 401s. Refreshing the user when that happens throws the aforementioned exception, rather than resulting in a valid cognito user with a new access token.

It's worth pointing out that even with a solution to this problem (or a workaround related to messing with expiration time on the tokens object), I still run into another issue related to the refresh action clearing the refresh token instead of leaving it alone or updating it. This bug has already been mentioned, just worth pointing out that fixing this bug and that bug together are required for a real refresh system:
https://github.com/aws/aws-sdk-net/issues/778

## Your Environment

* AWSSDK.Core version used: 3.3.19.0
* Service assembly and version used: 
* Operating System and version: Windows 10 1709 (OS Build 19299.192)
* Visual Studio version: Community 2017 15.5.4
* Targeted .NET platform: Core 2.1

## .NET Core Info

* .NET Core version used for development: 2.1.4
* .NET Core version installed in the environment where application runs: 2.1.4
* Output of `dotnet --info`: 
```
.NET Command Line Tools (2.1.4)

Product Information:
 Version:            2.1.4
 Commit SHA-1 hash:  5e8add2190

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.16299
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.1.4\

Microsoft .NET Core Shared Framework Host

  Version  : 2.0.5
  Build    : 17373eb129b3b05aa18ece963f8795d65ef8ea54
```
* Contents of project.json/project.csproj: 
project.json:
```
{
  "dependencies": {
    "AWSSDK.CognitoIdentity": "3.3.2.17",
    "AWSSDK.CognitoIdentityProvider": "3.3.8",
    "AWSSDK.Extensions.CognitoAuthentication": "0.9.1",
    "Microsoft.NETCore.UniversalWindowsPlatform": "5.2.2",
    "Microsoft.Toolkit.Uwp.Notifications": "2.1.0",
    "NETStandard.Library": "2.0.1",
    "Newtonsoft.Json": "10.0.3",
    "QueryString.NET": "1.0.0",
    "System.Net.Http": "4.3.3",
    "System.Threading": "4.3.0",
    "System.Threading.Tasks": "4.3.0",
    "System.Threading.Timer": "4.3.0",
    "Xamarin.Forms": "2.5.0.121934"
  },
  "frameworks": {
    "uap10.0.16299": {}
  },
  "runtimes": {
    "win10-arm": {},
    "win10-arm-aot": {},
    "win10-x86": {},
    "win10-x86-aot": {},
    "win10-x64": {},
    "win10-x64-aot": {}
  }
}
```
SDPNativeClient.UWP.csproj:
```
<?xml version="1.0" encoding="utf-8"?>
<Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
  <PropertyGroup>
    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
    <Platform Condition=" '$(Platform)' == '' ">x86</Platform>
    <ProjectGuid>{FD00B0CC-4454-440C-9F0E-00BFC3E99032}</ProjectGuid>
    <OutputType>AppContainerExe</OutputType>
    <AppDesignerFolder>Properties</AppDesignerFolder>
    <RootNamespace>SDPNativeClient.UWP</RootNamespace>
    <AssemblyName>SDPNativeClient.UWP</AssemblyName>
    <DefaultLanguage>en-US</DefaultLanguage>
    <TargetPlatformIdentifier>UAP</TargetPlatformIdentifier>
    <TargetPlatformVersion>10.0.16299.0</TargetPlatformVersion>
    <TargetPlatformMinVersion>10.0.16299.0</TargetPlatformMinVersion>
    <MinimumVisualStudioVersion>14</MinimumVisualStudioVersion>
    <EnableDotNetNativeCompatibleProfile>true</EnableDotNetNativeCompatibleProfile>
    <FileAlignment>512</FileAlignment>
    <ProjectTypeGuids>{A5A43C5B-DE2A-4C0C-9213-0A381AF9435A};{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}</ProjectTypeGuids>
    <PackageCertificateKeyFile>impulse.pfx</PackageCertificateKeyFile>
    <RuntimeIdentifiers>win10-arm;win10-arm-aot;win10-x86;win10-x86-aot;win10-x64;win10-x64-aot</RuntimeIdentifiers>
    <AppxAutoIncrementPackageRevision>True</AppxAutoIncrementPackageRevision>
    <AppxBundle>Always</AppxBundle>
    <AppxBundlePlatforms>x64</AppxBundlePlatforms>
    <PackageCertificateThumbprint>5B78E478072A4493A14DEF9F91A15E82DE7E1127</PackageCertificateThumbprint>
    <AppxSymbolPackageEnabled>False</AppxSymbolPackageEnabled>
    <AppxPackageDir>C:\Users\IEUser\Desktop\installables\</AppxPackageDir>
  </PropertyGroup>
  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Debug|ARM'">
    <DebugSymbols>true</DebugSymbols>
    <OutputPath>bin\ARM\Debug\</OutputPath>
    <DefineConstants>DEBUG;TRACE;NETFX_CORE;WINDOWS_UWP</DefineConstants>
    <NoWarn>;2008</NoWarn>
    <DebugType>full</DebugType>
    <PlatformTarget>ARM</PlatformTarget>
    <UseVSHostingProcess>false</UseVSHostingProcess>
    <ErrorReport>prompt</ErrorReport>
    <Prefer32Bit>true</Prefer32Bit>
  </PropertyGroup>
  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Release|ARM'">
    <OutputPath>bin\ARM\Release\</OutputPath>
    <DefineConstants>TRACE;NETFX_CORE;WINDOWS_UWP</DefineConstants>
    <Optimize>true</Optimize>
    <NoWarn>;2008</NoWarn>
    <DebugType>pdbonly</DebugType>
    <PlatformTarget>ARM</PlatformTarget>
    <UseVSHostingProcess>false</UseVSHostingProcess>
    <ErrorReport>prompt</ErrorReport>
    <Prefer32Bit>true</Prefer32Bit>
    <UseDotNetNativeToolchain>true</UseDotNetNativeToolchain>
  </PropertyGroup>
  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Debug|x64'">
    <DebugSymbols>true</DebugSymbols>
    <OutputPath>bin\x64\Debug\</OutputPath>
    <DefineConstants>DEBUG;TRACE;NETFX_CORE;WINDOWS_UWP</DefineConstants>
    <NoWarn>;2008</NoWarn>
    <DebugType>full</DebugType>
    <PlatformTarget>x64</PlatformTarget>
    <UseVSHostingProcess>false</UseVSHostingProcess>
    <ErrorReport>prompt</ErrorReport>
    <Prefer32Bit>true</Prefer32Bit>
    <UseDotNetNativeToolchain>false</UseDotNetNativeToolchain>
    <Optimize>false</Optimize>
  </PropertyGroup>
  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Release|x64'">
    <OutputPath>bin\x64\Release\</OutputPath>
    <DefineConstants>TRACE;NETFX_CORE;WINDOWS_UWP</DefineConstants>
    <Optimize>true</Optimize>
    <NoWarn>;2008</NoWarn>
    <DebugType>pdbonly</DebugType>
    <PlatformTarget>x64</PlatformTarget>
    <UseVSHostingProcess>false</UseVSHostingProcess>
    <ErrorReport>prompt</ErrorReport>
    <Prefer32Bit>true</Prefer32Bit>
    <UseDotNetNativeToolchain>false</UseDotNetNativeToolchain>
  </PropertyGroup>
  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Debug|x86'">
    <DebugSymbols>true</DebugSymbols>
    <OutputPath>bin\x86\Debug\</OutputPath>
    <DefineConstants>DEBUG;TRACE;NETFX_CORE;WINDOWS_UWP</DefineConstants>
    <NoWarn>;2008</NoWarn>
    <DebugType>full</DebugType>
    <PlatformTarget>x86</PlatformTarget>
    <UseVSHostingProcess>false</UseVSHostingProcess>
    <ErrorReport>prompt</ErrorReport>
    <Prefer32Bit>true</Prefer32Bit>
  </PropertyGroup>
  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Release|x86'">
    <OutputPath>bin\x86\Release\</OutputPath>
    <DefineConstants>TRACE;NETFX_CORE;WINDOWS_UWP</DefineConstants>
    <Optimize>true</Optimize>
    <NoWarn>;2008</NoWarn>
    <DebugType>pdbonly</DebugType>
    <PlatformTarget>x86</PlatformTarget>
    <UseVSHostingProcess>false</UseVSHostingProcess>
    <ErrorReport>prompt</ErrorReport>
    <Prefer32Bit>true</Prefer32Bit>
    <UseDotNetNativeToolchain>true</UseDotNetNativeToolchain>
  </PropertyGroup>
  <ItemGroup>
    <Compile Include="API\IPCService.cs" />
    <Compile Include="Platform\VPNService.cs" />
    <Compile Include="App.xaml.cs">
      <DependentUpon>App.xaml</DependentUpon>
    </Compile>
    <Compile Include="MainPage.xaml.cs">
      <DependentUpon>MainPage.xaml</DependentUpon>
    </Compile>
    <Compile Include="OpenAppService.cs" />
    <Compile Include="Platform\PlatformEffectUWP.cs" />
    <Compile Include="Platform\PlatformReplaceAppTitle.cs" />
    <Compile Include="Properties\AssemblyInfo.cs" />
    <Compile Include="Toaster.cs" />
  </ItemGroup>
  <ItemGroup>
    <AppxManifest Include="Package.appxmanifest">
      <SubType>Designer</SubType>
    </AppxManifest>
    <None Include="impulse.pfx">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
    <None Include="project.json" />
    <None Include="SDPNativeClient.UWP_TemporaryKey.pfx" />
  </ItemGroup>
  <ItemGroup>
    <Content Include="app_icon.png" />
    <Content Include="Assets\Square44x44Logo.targetsize-16.png" />
    <Content Include="Assets\Square44x44Logo.targetsize-24.png" />
    <Content Include="Assets\Square44x44Logo.targetsize-256.png" />
    <Content Include="Assets\Square44x44Logo.targetsize-32.png" />
    <Content Include="Assets\Square44x44Logo.targetsize-48.png" />
    <Content Include="Bundle.Mapping.txt" />
    <Content Include="libfko.arm.dll" />
    <Content Include="libfko.x86.dll" />
    <Content Include="libfko.x64.dll" />
    <Content Include="menu_1.png" />
    <Content Include="Properties\Default.rd.xml" />
    <Content Include="Assets\LockScreenLogo.scale-100.png" />
    <Content Include="Assets\LockScreenLogo.scale-125.png" />
    <Content Include="Assets\LockScreenLogo.scale-150.png" />
    <Content Include="Assets\LockScreenLogo.scale-200.png" />
    <Content Include="Assets\LockScreenLogo.scale-400.png" />
    <Content Include="Assets\SplashScreen.scale-100.png" />
    <Content Include="Assets\SplashScreen.scale-125.png" />
    <Content Include="Assets\SplashScreen.scale-150.png" />
    <Content Include="Assets\SplashScreen.scale-200.png" />
    <Content Include="Assets\SplashScreen.scale-400.png" />
    <Content Include="Assets\Square150x150Logo.scale-200.png" />
    <Content Include="Assets\Square44x44Logo.scale-100.png" />
    <Content Include="Assets\Square44x44Logo.scale-125.png" />
    <Content Include="Assets\Square44x44Logo.scale-150.png" />
    <Content Include="Assets\Square44x44Logo.scale-200.png" />
    <Content Include="Assets\Square44x44Logo.scale-400.png" />
    <Content Include="Assets\Square44x44Logo.targetsize-16_altform-unplated.png" />
    <Content Include="Assets\Square44x44Logo.targetsize-24_altform-unplated.png" />
    <Content Include="Assets\Square44x44Logo.targetsize-32_altform-unplated.png" />
    <Content Include="Assets\Square44x44Logo.targetsize-48_altform-unplated.png" />
    <Content Include="Assets\Square44x44Logo.targetsize-256_altform-unplated.png" />
    <Content Include="Assets\StoreLogo.png" />
    <Content Include="Assets\Wide310x150Logo.scale-100.png" />
    <Content Include="Assets\Wide310x150Logo.scale-125.png" />
    <Content Include="Assets\Wide310x150Logo.scale-150.png" />
    <Content Include="Assets\Wide310x150Logo.scale-200.png" />
    <Content Include="Assets\Wide310x150Logo.scale-400.png" />
    <Content Include="WindowsFullTrustProcess.exe">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </Content>
    <Content Include="refresh.png" />
  </ItemGroup>
  <ItemGroup>
    <ApplicationDefinition Include="App.xaml">
      <Generator>MSBuild:Compile</Generator>
      <SubType>Designer</SubType>
    </ApplicationDefinition>
    <Page Include="MainPage.xaml">
      <Generator>MSBuild:Compile</Generator>
      <SubType>Designer</SubType>
    </Page>
  </ItemGroup>
  <ItemGroup>
    <PackageReference Include="AWSSDK.CognitoIdentity">
      <Version>3.3.2.17</Version>
    </PackageReference>
    <PackageReference Include="AWSSDK.CognitoIdentityProvider">
      <Version>3.3.8</Version>
    </PackageReference>
    <PackageReference Include="AWSSDK.Extensions.CognitoAuthentication">
      <Version>0.9.1</Version>
    </PackageReference>
    <PackageReference Include="BouncyCastle">
      <Version>1.8.1</Version>
    </PackageReference>
    <PackageReference Include="Fwknop">
      <Version>1.0.1</Version>
    </PackageReference>
    <PackageReference Include="Microsoft.NETCore.UniversalWindowsPlatform">
      <Version>6.0.6</Version>
    </PackageReference>
    <PackageReference Include="Microsoft.Toolkit.Uwp.Notifications">
      <Version>2.1.1</Version>
    </PackageReference>
    <PackageReference Include="NETStandard.Library">
      <Version>2.0.1</Version>
    </PackageReference>
    <PackageReference Include="Newtonsoft.Json">
      <Version>11.0.1-beta3</Version>
    </PackageReference>
    <PackageReference Include="QueryString.NET">
      <Version>1.0.0</Version>
    </PackageReference>
    <PackageReference Include="Serilog">
      <Version>2.6.1-dev-00938</Version>
    </PackageReference>
    <PackageReference Include="Serilog.Sinks.Console">
      <Version>3.1.1</Version>
    </PackageReference>
    <PackageReference Include="Serilog.Sinks.Debug">
      <Version>1.0.1-dev-00015</Version>
    </PackageReference>
    <PackageReference Include="Serilog.Sinks.File">
      <Version>4.0.1-dev-00790</Version>
    </PackageReference>
    <PackageReference Include="System.Net.Http">
      <Version>4.3.3</Version>
    </PackageReference>
    <PackageReference Include="System.Security.Cryptography.OpenSsl">
      <Version>4.4.0</Version>
    </PackageReference>
    <PackageReference Include="System.Threading">
      <Version>4.3.0</Version>
    </PackageReference>
    <PackageReference Include="System.Threading.Tasks">
      <Version>4.3.0</Version>
    </PackageReference>
    <PackageReference Include="System.Threading.Timer">
      <Version>4.3.0</Version>
    </PackageReference>
    <PackageReference Include="Xamarin.Forms">
      <Version>2.5.0.122203</Version>
    </PackageReference>
  </ItemGroup>
  <ItemGroup>
    <SDKReference Include="WindowsDesktop, Version=10.0.16299.0">
      <Name>Windows Desktop Extensions for the UWP</Name>
    </SDKReference>
  </ItemGroup>
  <ItemGroup>
    <WCFMetadata Include="Connected Services\" />
  </ItemGroup>
  <PropertyGroup Condition=" '$(VisualStudioVersion)' == '' or '$(VisualStudioVersion)' &lt; '14.0' ">
    <VisualStudioVersion>14.0</VisualStudioVersion>
  </PropertyGroup>
  <Import Project="..\SDPNativeClient\SDPNativeClient.projitems" Label="Shared" Condition="Exists('..\SDPNativeClient\SDPNativeClient.projitems')" />
  <Import Project="..\..\WindowsAppServiceIPC\WindowsAppServiceIPC.projitems" Label="Shared" />
  <Import Project="..\..\WindowsFullTrustProcessShared\WindowsFullTrustProcessShared.projitems" Label="Shared" />
  <Import Project="$(MSBuildExtensionsPath)\Microsoft\WindowsXaml\v$(VisualStudioVersion)\Microsoft.Windows.UI.Xaml.CSharp.targets" />
  <PropertyGroup>
    <PreBuildEvent>"$(msbuildbinpath)\msbuild.exe" $(SolutionPath) /t:WindowsFullTrustProcess:Rebuild</PreBuildEvent>
  </PropertyGroup>
  <PropertyGroup>
    <PostBuildEvent>
    </PostBuildEvent>
  </PropertyGroup>
  <!-- To modify your build process, add your task inside one of the targets below and uncomment it. 
       Other similar extension points exist, see Microsoft.Common.targets.
  <Target Name="BeforeBuild">
  </Target>
  <Target Name="AfterBuild">
  </Target>
  -->
</Project>
```

[logging]: http://docs.aws.amazon.com/sdk-for-net/v3/developer-guide/net-dg-config-other.html

**Comment by peachwyss** [comment]
@jeffreythomasprice I have the same issue, using DateTime.Now.AddHours(1)) seems to work-around the issue. But I also think this is a bug.

**Comment by RidhwanDev** [comment]
Yeah, After a few months of having this Issue I think I know why it hasn't been fixed yet.  It seems as though it was intended to work this way but the issue comes with a badly named parameter. 

The reason DateTime.Now.AddHours(1) works is because that is doing the check against expiration time. I'm guessing we've all written something to check if the users 30 days are over before even getting to this function call so we will never make the call to refresh the token if their refresh token should be expired anyway. 

public CognitoUserSession(string idToken, string accessToken, string refreshToken, DateTime issuedTime, DateTime expirationTime);

We're passing expirationTime as the token expiration time. However when refreshing the token, **I think this should be passing in the Expiration Time of the RefreshToken**.

I've changed this
```
user.SessionTokens = new CognitoUserSession(
	user.SessionTokens.IdToken,
	user.SessionTokens.AccessToken,
	user.SessionTokens.RefreshToken,
	user.SessionTokens.IssuedTime,
	session.TokenExpiresServer
);
```

into this
```
user.SessionTokens = new CognitoUserSession(
	user.SessionTokens.IdToken,
	user.SessionTokens.AccessToken,
	user.SessionTokens.RefreshToken,
	user.SessionTokens.IssuedTime,
        user.session.RefreshTokenExpiresServer
);
```

My RefreshTokenExpiresServer is set to be 25 days from when the user gets authenticated. This is what I think is the issue here is anyway. @jeffreythomasprice  @sstevenkang @curtisshipley 

This has fixed the issue for me.

**Comment by RidhwanDev** [comment]
```
   private void EnsureUserAuthenticated()
        {
            if (SessionTokens == null || !SessionTokens.IsValid())
            {
                throw new NotAuthorizedException("User is not authenticated.");
            }
        }
```

```
public CognitoUserSession SessionTokens
{
......
        public bool IsValid()
             {
                    DateTime currentTimeStamp = DateTime.Now;

                     return (currentTimeStamp.CompareTo(ExpirationTime) < 0);
             }
}
```

The 'expirationTime' we pass in is literally compared to the current time before we even it even attempts to authenticate.

It just needs to be the expirationTime for the RefreshToken.

**Comment by assyadh** [comment]
Hi everyone,

I'll have a look at this shortly and will narrow down the issue. 

H

**Comment by ddiimmiittaarr** [comment]
@ assyadh 

Awesome!!!
I took like more than  a year !!!!

Finger crossed for 2021

**Comment by darbio** [comment]
I am also struggling with this.

From my understanding, we should be able to pass a refresh token into the service and nothing else (we don't need a user id, access token, id token etc.). The refresh token should be authenticated and return new tokens. This is how other ID providers work.

**Comment by motoko89** [comment]
This is living proof the AWS loves the .NET SDK more than any other languages! Probably will be fixed in 3 years

**Comment by philipogorman** [comment]
The `DateTime.Now.AddHours(1) ` work around seems to work

**Comment by michuny25** [comment]
Following workaround worked for me, full code:
`public async Task<LoginModel> RefreshToken(string userId, string idToken, string refreshToken)
        {
            var user = new Amazon.Extensions.CognitoAuthentication.CognitoUser(userId, _appClientId, _userPool, _identityProvider);

            user.SessionTokens = new CognitoUserSession(idToken, string.Empty, refreshToken, DateTime.UtcNow,
            DateTime.UtcNow.AddDays(1));

            var authResponse = await user.StartWithRefreshTokenAuthAsync(new InitiateRefreshTokenAuthRequest()
            {
                AuthFlowType = AuthFlowType.REFRESH_TOKEN
            });

            return new LoginModel
            {
                IdToken = authResponse.AuthenticationResult.IdToken,
                AccessToken = authResponse.AuthenticationResult.AccessToken,
                RefreshToken = authResponse.AuthenticationResult.RefreshToken
            };
        }`

**Comment by atiyka** [comment]
Hi all, I'm also struggling with this issue.
If I renew the tokens before the access token expires, it works fine, but as soon as the token expires I can't renew them - the `StartWithRefreshTokenAuthAsync()` throws an exception that the `User is not authenticated.`
I tried this workaround to set the expiration time to a future date (`DateTime.Now.AddHours(1)`). The same thing happens: if the access token isn't expired it works fine, but if it is, it throws an exception that: `Refresh Token has been revoked`.
Any suggestions?

**Comment by RidhwanDev** [comment]
I've explained this before, the reason `DateTime.Now.AddHours(1)` or `DateTime.UtcNow.AddDays(1)` works is because it's the property ExpirationTime actually means RefreshTokenExpiration time. If you look at the code and how it is used, you will see that it is treated as a refresh token expiration. They have just not named it properly or made it clear enough that it is supposed to be that. 

In my code I set this to 25 days from when the user first logs in. So if you logged in today, you have 25 more days to log in again and I will just refresh your token and give you a new one, and reset your 25 days. However if you log in after this 25 days then sorry mate your creds are expired and you need to log in again. 

So everybody putting `DateTime.Now.AddHours(1)` in is just saying that I will always refresh your token because your refresh token is always valid until 1 hour in the future. 

Does that make sense?

**Comment by atiyka** [comment]
Yeah, it makes sens, thank you, I'll check it

**Comment by philipogorman** [comment]
Here is the code we use to refresh: 
`
public async Task<SignInContext> RefreshToken(string idToken, string accessToken, String refreshToken, DateTime issued, DateTime expires)
        {
            try
            {
                var provider = new AmazonCognitoIdentityProviderClient(new AnonymousAWSCredentials(), RegionEndpoint);

                CognitoUserPool userPool = new CognitoUserPool(PoolId, ClientId, provider);
                CognitoUser user = new CognitoUser("", ClientId, userPool, provider);

                // The  "Refresh token expiration (days)" (Cognito->UserPool->General Settings->App clients->Show Details) is the
                // amount of time since the last login that you can use the refresh token to get new tokens.
                // After that period the refresh will fail
                // Using DateTime.Now.AddHours(1) is a workaround for https://github.com/aws/aws-sdk-net-extensions-cognito/issues/24
                user.SessionTokens = new CognitoUserSession(idToken, accessToken, refreshToken, issued, DateTime.Now.AddHours(1));

                AuthFlowResponse context = await user.StartWithRefreshTokenAuthAsync(new InitiateRefreshTokenAuthRequest
                {
                    AuthFlowType = AuthFlowType.REFRESH_TOKEN_AUTH
                })
                .ConfigureAwait(false);

                return new SignInContext(CognitoResult.Ok)
                {
                    IdToken = context.AuthenticationResult?.IdToken,
                    AccessToken = context.AuthenticationResult?.AccessToken,
                    TokenIssued = user.SessionTokens.IssuedTime,
                    Expires = user.SessionTokens.ExpirationTime,
                    SessionId = context.SessionID
                };
            }
            catch (NotAuthorizedException ne)
            {
                //https://docs.aws.amazon.com/cognito/latest/developerguide/amazon-cognito-user-pools-using-tokens-with-identity-providers.html
                // refresh tokens will expire - user must login manually every x days (see user pool -> app clients -> details)
                return new SignInContext(CognitoResult.RefreshNotAuthorized) { ResultMessage = ne.Message };
            }
            catch (WebException webEx)
            {
                // we get a web exception when we cant connect to aws - means we are offline
                return new SignInContext(CognitoResult.NoInternetConection) { ResultMessage = webEx.Message };
            }
            catch (Exception ex)
            {
                // cant find the full list of possible execptions in aws docs
                return new SignInContext(CognitoResult.Unknown) { ResultMessage = ex.Message };
            }
        }
`

**Comment by atiyka** [comment]
Thanks for the answers!
I'm doing the same thing. So, when I login with user and password it gives me the access tokens, which expires within one hour. If I call that refresh function `StartWithRefreshTokenAuthAsync` **within that _one_ hour** it gives me another tokens (except refresh token) which are valid for one hour from that point, and so on. It doesn't matter what I set as expiration.
**But**, if that specific one hour **passes**, without refreshing any tokens it doesn't gives me new tokens. So if try to call `StartWithRefreshTokenAuthAsync()` after more than one hour since the last login/refresh, it throws the exception what I already posted (`Refresh Token has been revoked`)...

Are you refreshing the tokens often, or is this scenario possible for you, to leave the app without communication with the server for more than one hour?

**Comment by philipogorman** [comment]
there's a difference between the "access token" and the refresh token.
The access token must be refreshed every hour in my case - we set a timer in the app to do that. This is the token that is used in the api calls.

The refresh token, is the token used to refresh the access token. The  "Refresh token expiration (days)" (Cognito->UserPool->General Settings->App clients->Show Details) is the amount of time since the last login that you can use the refresh token to get new tokens. After that period the refresh will fail.

It took a lot of fumbling around to get this working as desired!

**Comment by NGL321** [comment]
Im not sure who to ask, but @philipogorman and @RidhwanDev you seem to be up to date (Hamza no longer works on the SDK team). Is this a bug still being experienced? 
If not, I want to leave a comment with official guidance and close it, but if it is I want to make sure it gets escalated ASAP.

😸 😷

**Comment by philipogorman** [comment]
yes - we still have the workaround in our code, I haven't tested without the workaround though:

`// The  "Refresh token expiration (days)" (Cognito->UserPool->General Settings->App clients->Show Details) is the
            // amount of time since the last login that you can use the refresh token to get new tokens.
            // After that period the refresh will fail
            // Using DateTime.Now.AddHours(1) is a workaround for https://github.com/aws/aws-sdk-net-extensions-cognito/issues/24
            user.SessionTokens = new CognitoUserSession(idToken, accessToken, refreshToken, issued, DateTime.Now.AddHours(1));
`

**Comment by RidhwanDev** [comment]
@philipogorman I think in your previous comment you've also figured out the design behind it however you're still using the `DateTime.Now.AddHours(1)` work around. This work around means you will **always** refresh the user's token. At the point in which you authenticate the user, you should store a property with a name like `RefreshTokenExpirationTime` and set this anywhere between an hour to 30 days from that moment you authenticate them. This `RefreshTokenExpirationTime` will dictate how long they have before their refresh token expires. I've always set mine to 25 days from the moment they are authenticated, which means if they keep logging in within 25 days they will always remained signed in -- they will be authenticated again and their refresh token will have a new expiration time. However if they login again 26 days since their last sign in then they will need to sign in again.

Using `DateTime.Now.AddHours(1)` will work because cognito will always try to refresh their user's token. It's essentially a **force** refresh. However **it will not work in this situation**: If the user has signed in and then leaves and doesn't come back to the app for over 30 days. (At this point the actual refresh token be expired, unless you have changed the expiration time of your refresh tokens) Your code of `DateTime.Now.AddHours(1)` will try to force refreshing the token again which will fail. Your app may or may not handle this gracefully but it certainly isn't the behaviour you want.

@NGL321 I'm not sure if my comments have made sense, people don't seem to understand it. I don't think theres anything wrong with the code here per se but argument name is very misleading and there should be proper instructions on how to use it.

**Comment by philipogorman** [comment]
@RidhwanDev  your summary sounds correct to me

**Comment by ashishdhingra** [comment]
Hi @jeffreythomasprice,

Please have a look at the comments made by @philipogorman and @RidhwanDev, and advise if the issue could be closed.

Thanks,
Ashish

**Comment by TroySchmidt** [comment]
This is still a "bug".  I think the issue is that most people initiate the login and get the expiry for the access token.  Then they try and use the same set of session tokens to refresh.  The code is checking the expiry date and failing.  The issue is that between authenticating and refreshing there are two different time frames, but only one is stored in the session tokens.  So you have to provide your own new expiry time on the refresh that is valid so the SDK expiry validation passes.  This usually isn't an issue since that is usually a month so knowing the exact time table for that is not needed.  Also considering a new refresh token is given then that almost always works and doesn't require a login with password again.

**Comment by BatteryAcid** [comment]
What’s the difference between the the `ExpirationTime` set in an `CognitoUserSession` object and the configuration set in the AWS Console under my User Pool app client’s “Refresh token expiration”?

For the `CognitoUserSession`, setting the ExpirationTime to 1 hour will expire the Refresh token after an hour of no refresh.  In my User Pool for the App client, the refresh token is set to 30 days.  So, I’m assuming that whatever I set in the `CognitoUserSession` for the refresh ExpirationTime takes precedence over the User Pool configuration?  

Should I be matching the expiration time of what’s configured under the App client’s refresh token expiration to what I’m setting in `CognitoUserSession`? Or is this just to allow flexibility in how my app wants to handle the refresh?

I’m not sure how these two settings work together.

**Comment by davidsaldana** [comment]
In case anyone is still seeing this, the issuedTime and expirationTime parameters don't really matter for this call at all. I set these values to DateTime.MinValue and the code works just fine. See below:
`cognitoUser.SessionTokens = new CognitoUserSession(
                        credentials.IdTokenForRefresh,
                        credentials.AccessTokenForRefresh,
                        credentials.RefreshTokenForRefresh,
                        // issuedTime and expiration time don't matter here
                        // References:
                        //  https://github.com/aws/aws-sdk-net-extensions-cognito/issues/24
                        issuedTime: DateTime.MinValue,
                        expirationTime: DateTime.MinValue);`
I've tested with an expired refresh token and an unexpired refresh token. In both cases, the call is attempted normally. However, they differ in that the call with the expired token returns a "Refresh Token has expired" response error and the call with the unexpired refresh token gets you a refreshed set of id and access tokens.

In my opinion what the aws cognito developers really should've done is simply provide an overloaded constructor like the following:
`public CognitoUserSession(string idToken, string accessToken, string refreshToken)
        {
            IdToken = idToken;
            AccessToken = accessToken;
            RefreshToken = refreshToken;
        }`
This would've prevented lots of confusion. Well that and better documentation. The aws sdk is certainly not the friendliest for developers.
